File: packages/e2e-test/src/__tests__/dependency.test.ts
/**
 * End-to-End Tests for dependency resolution and ordering.
 *
 * These tests verify that the CLI correctly:
 * - Handles successful dependency resolution
 * - Detects missing dependency providers
 * - Detects ambiguous dependencies (multiple providers)
 * - Detects circular dependencies
 * - Handles platform-specific dependency availability
 */
import { describe, expect, it } from "bun:test";
// oxlint-disable-next-line import/no-unassigned-import
import "@dotfiles/testing-helpers";
import { Architecture, Platform } from "@dotfiles/core";
import { withMockServer } from "./helpers/mock-server";
import { TestHarness } from "./helpers/TestHarness";
import type { ITestTarget } from "./helpers/types";

describe("E2E: dependency resolution", () => {
  // Dependency tests use manual installers, no mock endpoints needed
  withMockServer();

  const platformConfigs: ReadonlyArray<ITestTarget> = [
    { platform: Platform.MacOS, architecture: Architecture.Arm64, name: "macOS ARM64" },
    { platform: Platform.Linux, architecture: Architecture.X86_64, name: "Linux x86_64" },
  ];

  for (const config of platformConfigs) {
    describe(`${config.name}`, () => {
      describe("dependency resolution", () => {
        it("generates successfully when dependencies are satisfied", async () => {
          const harness = new TestHarness({
            testDir: import.meta.dir,
            configPath: "fixtures/dependency-success/config.ts",
            platform: config.platform,
            architecture: config.architecture,
            cleanBeforeRun: true,
          });
          const result = await harness.generate();
          expect(result.code).toBe(0);
        });

        it("fails when a dependency provider is missing", async () => {
          const harness = new TestHarness({
            testDir: import.meta.dir,
            configPath: "fixtures/dependency-missing/config.ts",
            platform: config.platform,
            architecture: config.architecture,
            cleanBeforeRun: true,
          });
          const result = await harness.generate();
          expect(result.code).toBe(1);
          const combinedOutput = `${result.stdout}${result.stderr}`;
          expect(combinedOutput).toContain('Missing dependency: tool "dependency-consumer-missing" requires binary');
          expect(combinedOutput).toContain("missing-provider");
        });

        it("fails when multiple tools provide the same dependency", async () => {
          const harness = new TestHarness({
            testDir: import.meta.dir,
            configPath: "fixtures/dependency-ambiguous/config.ts",
            platform: config.platform,
            architecture: config.architecture,
            cleanBeforeRun: true,
          });
          const result = await harness.generate();
          expect(result.code).toBe(1);
          const combinedOutput = `${result.stdout}${result.stderr}`;
          expect(combinedOutput).toContain(
            'Ambiguous dependency: binary "shared-dependency" is provided by multiple tools',
          );
          expect(combinedOutput).toContain("dependency-provider-a");
          expect(combinedOutput).toContain("dependency-provider-b");
          expect(combinedOutput).toContain("dependency-consumer-ambiguous");
        });

        it("fails when dependencies create a cycle", async () => {
          const harness = new TestHarness({
            testDir: import.meta.dir,
            configPath: "fixtures/dependency-circular/config.ts",
            platform: config.platform,
            architecture: config.architecture,
            cleanBeforeRun: true,
          });
          const result = await harness.generate();
          expect(result.code).toBe(1);
          const combinedOutput = `${result.stdout}${result.stderr}`;
          expect(combinedOutput).toContain("Circular dependency detected between tools");
          expect(combinedOutput).toContain("dependency-cycle-a");
          expect(combinedOutput).toContain("dependency-cycle-b");
        });

        it("fails when the dependency provider is unavailable on the active platform", async () => {
          const harness = new TestHarness({
            testDir: import.meta.dir,
            configPath: "fixtures/dependency-platform-mismatch/config.ts",
            platform: config.platform,
            architecture: config.architecture,
            cleanBeforeRun: true,
          });
          const result = await harness.generate();
          expect(result.code).toBe(1);
          const combinedOutput = `${result.stdout}${result.stderr}`;
          expect(combinedOutput).toContain('Missing dependency: tool "dependency-platform-consumer" requires binary');
          expect(combinedOutput).toContain("platform-specific-binary");
        });
      });
    });
  }
});

File: tests/e2e/dependency_test.go
package e2e

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestE2EDependencyResolution(t *testing.T) {
	t.Parallel()

	t.Run("generates successfully when dependencies are satisfied", func(t *testing.T) {
		h := NewTestHarness(t, HarnessOptions{
			ConfigPath: "config.ts",
		})
		h.CopyFixture("dependency-success")

		_, _, exitCode, err := h.Generate("-d")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if exitCode != 0 {
			t.Fatalf("expected generate to succeed, but got exit code %d", exitCode)
		}
	})

	t.Run("fails when a dependency provider is missing", func(t *testing.T) {
		h := NewTestHarness(t, HarnessOptions{
			ConfigPath: "config.ts",
		})
		h.CopyFixture("dependency-missing")

		stdout, stderr, exitCode, err := h.Generate("-d")
		if err == nil && exitCode == 0 {
			t.Fatalf("expected generate to fail, but succeeded")
		}

		output := stdout + stderr
		if !strings.Contains(output, "depends on missing dependency") && !strings.Contains(output, "requires binary") {
			t.Fatalf("expected output to mention missing dependency, but got:\n%s", output)
		}
		if !strings.Contains(output, "missing-provider") {
			t.Fatalf("expected output to mention missing-provider, but got:\n%s", output)
		}
	})

	t.Run("fails when multiple tools provide the same dependency", func(t *testing.T) {
		h := NewTestHarness(t, HarnessOptions{
			ConfigPath: "config.ts",
		})
		h.CopyFixture("dependency-ambiguous")

		stdout, stderr, exitCode, err := h.Generate("-d")
		if err == nil && exitCode == 0 {
			t.Fatalf("expected generate to fail, but succeeded")
		}

		output := stdout + stderr
		if !strings.Contains(output, "ambiguous dependency") && !strings.Contains(output, "provided by multiple tools") {
			t.Fatalf("expected output to mention ambiguous dependency, but got:\n%s", output)
		}
		if !strings.Contains(output, "shared-dependency") {
			t.Fatalf("expected output to mention shared-dependency, but got:\n%s", output)
		}
	})

	t.Run("fails when dependencies create a cycle", func(t *testing.T) {
		h := NewTestHarness(t, HarnessOptions{
			ConfigPath: "config.ts",
		})
		h.CopyFixture("dependency-circular")

		stdout, stderr, exitCode, err := h.Generate("-d")
		if err == nil && exitCode == 0 {
			t.Fatalf("expected generate to fail, but succeeded")
		}

		output := stdout + stderr
		if !strings.Contains(output, "cycle") && !strings.Contains(output, "detected") {
			t.Fatalf("expected output to mention circular dependency, but got:\n%s", output)
		}
		if !strings.Contains(output, "dependency-cycle-a") || !strings.Contains(output, "dependency-cycle-b") {
			t.Fatalf("expected output to mention cycle members, but got:\n%s", output)
		}
	})

	t.Run("fails when the dependency provider is unavailable on the active platform", func(t *testing.T) {
		h := NewTestHarness(t, HarnessOptions{
			ConfigPath: "config.ts",
		})
		h.CopyFixture("dependency-platform-mismatch")

		stdout, stderr, exitCode, err := h.Generate("-d")
		if err == nil && exitCode == 0 {
			t.Fatalf("expected generate to fail, but succeeded")
		}

		output := stdout + stderr
		if !strings.Contains(output, "depends on missing dependency") && !strings.Contains(output, "requires binary") {
			t.Fatalf("expected output to mention missing dependency, but got:\n%s", output)
		}
	})
}

// A dependency on a disabled tool's binary is a reference the generated bin-name
// registry accepts, so it must load: the run carries on without the disabled provider
// and generates every other tool, including ones with no part in the dependency.
func TestE2EDependencyOnDisabledProvider(t *testing.T) {
	t.Parallel()

	h := NewTestHarness(t, HarnessOptions{ConfigPath: "config.ts"})
	h.CopyFixture("dependency-disabled-provider")

	configContent := `export default {
  paths: {
    generatedDir: "` + filepath.ToSlash(filepath.Join(h.TempDir, ".generated")) + `",
    homeDir: "{paths.generatedDir}/user-home",
    targetDir: "{paths.generatedDir}/user-bin",
    toolConfigsDir: "` + filepath.ToSlash(filepath.Join(h.TempDir, "tools")) + `",
  },
};`
	if err := os.WriteFile(filepath.Join(h.TempDir, "config.ts"), []byte(configContent), 0644); err != nil {
		t.Fatalf("failed to write config.ts: %v", err)
	}

	stdout, stderr, exitCode, err := h.Generate()
	if err != nil || exitCode != 0 {
		t.Fatalf("expected generate to succeed, got exit code %d, err %v\nstdout: %s\nstderr: %s", exitCode, err, stdout, stderr)
	}

	output := stdout + stderr
	if !strings.Contains(output, `provided by disabled tool "provider"`) {
		t.Errorf("expected the run to name the disabled provider, but got:\n%s", output)
	}
	if strings.Contains(output, "missing dependency") {
		t.Errorf("expected the disabled provider not to be reported as missing, but got:\n%s", output)
	}

	h.AssertShimExistsAndExecutable("unrelated-bin")
	h.AssertShimExistsAndExecutable("consumer-bin")

	if _, err := os.Stat(filepath.Join(h.TempDir, ".generated", "user-bin", "provider-bin")); !os.IsNotExist(err) {
		t.Errorf("expected no shim for the disabled provider's binary (err=%v)", err)
	}
}
